aws/resource_aws_lambda_permission.go

```diff
@@ -313,47 +326,52 @@ func resourceAwsLambdaPermissionDelete(d *schema.ResourceData, meta interface{})
 		return err
 	}
 
+	params := &lambda.GetPolicyInput{
+		FunctionName: aws.String(d.Get("function_name").(string)),
+	}
+	if v, ok := d.GetOk("qualifier"); ok {
+		params.Qualifier = aws.String(v.(string))
+	}
+	var statement *LambdaPolicyStatement
 	err = resource.Retry(5*time.Minute, func() *resource.RetryError {
-		log.Printf("[DEBUG] Checking if Lambda permission %q is deleted", d.Id())
-
-		params := &lambda.GetPolicyInput{
-			FunctionName: aws.String(d.Get("function_name").(string)),
-		}
-		if v, ok := d.GetOk("qualifier"); ok {
-			params.Qualifier = aws.String(v.(string))
-		}
-
 		log.Printf("[DEBUG] Looking for Lambda permission: %s", *params)
 		resp, err := conn.GetPolicy(params)
 		if err != nil {
-			if awsErr, ok := err.(awserr.Error); ok {
-				if awsErr.Code() == "ResourceNotFoundException" {
-					return nil
-				}
+			if isAWSErr(err, "ResourceNotFoundException", "") {
+				return nil
 			}
 			return resource.NonRetryableError(err)
 		}
-
 		if resp.Policy == nil {
 			return nil
 		}
 
-		policyInBytes := []byte(*resp.Policy)
-		policy := LambdaPolicy{}
-		err = json.Unmarshal(policyInBytes, &policy)
+		statement, err = getLambdaPolicyStatement(resp, d.Id())
 		if err != nil {
-			return resource.RetryableError(
-				fmt.Errorf("Error unmarshalling Lambda policy: %s", err))
+			return nil
 		}
+		return nil
+	})
 
-		_, err = findLambdaPolicyStatementById(&policy, d.Id())
-		if err != nil {
+	if isResourceTimeoutError(err) {
+		resp, err := conn.GetPolicy(params)
```

Comment:
What happens if `err` is not `nil` and not `ResourceNotFoundException`? 😄 In general I would recommend setting up all error handling before response/output handling. 👍

Reply:
@bflad that is handled below (line 376 of the new changes) - does that need to be moved above the `if statement != nil` block?

Reply:
Since this is using `:=` instead of `=`, the `err` object is only scoped within the `if isResourceTimeoutError(err)` conditional, so only the original resource timeout error can reach line 376 😄